### PR TITLE
make_prg update PR series: 2. make_prg/utils refactoring/updates

### PR DESCRIPTION
**Pull request by leoisl**
Hello,

continuing the update PR series, now moving to `make_prg/utils` module. This module was created in this PR, before we had `seq_utils.py`, `io_utils.py` and `prg_encoder.py` inside the `make_prg` dir. With some refactoring and with quite a good amount of new code, the number of utils files increased from 3 to 7, so I decided to create an utils module. The annoying part of moving + refactoring though is that git sometimes does not cope well with these simultaneous changes, and for some files, e.g. `seq_utils.py`, it says `make_prg/seq_utils.py` was completely removed, and `make_prg/utils/seq_utils.py` is a completely new file, while it was just a file renaming and several changes. This makes reviewing a bit more laborious, but if you prefer we can improve on this (i.e. commit the file renaming operations to this branch, and then PR just the code changes).

Some points to help on review:

* Class `GFA_Output` was extracted from `make_prg/io_utils.py` and refactored into its own source file, `make_prg/utils/gfa.py` with very few changes;
* Many functions in `make_prg/utils/io_utils.py` are marked with a note that they are not unit tested, as it would be a bit complicated to unit test them. However, they are tested by integration tests. I think the only function that remained in this source file from the previous code is `load_alignment_file()`, all the others are new;
* The code in files `make_prg/utils/gfa.py` and `make_prg/utils/recursive_tree_drawer.py` are tested through a single big bang test (i.e. a single test that check if they produce the expected output). This is not ideal, and it is more of a regression test checking if we keep producing the same output over the next versions. However, I think it is fine, as these are not crucial outputs produced by `make_prg`: we don't often look at the `.gfa` files, mostly for debugging. The recursive tree drawer is exclusive for debugging, not for user consumption, to understand the recursive tree generated by `make_prg`, which is rather complicated to visualise in debug mode;
* `prg_encoder.py` has basically no changes.

Thanks a lot for the help!

### CODE REVIEW

File: make_prg/utils/io_utils.py
Comment (mbhall88) [inline make_prg/utils/io_utils.py:15]:
```suggestion
def load_alignment_file(msa_file: Union[str, Path], alignment_format: str) -> MSA:
```

I guess you'll need to add `Union` from `typing`

File: make_prg/utils/io_utils.py
Comment (mbhall88) [inline make_prg/utils/io_utils.py:49]:
```suggestion
    for f in files:
        fullpath = os.path.join(path, f)
        if os.path.isdir(fullpath):
            remove_empty_folders(fullpath)
```

File: make_prg/utils/io_utils.py
Comment (mbhall88) [inline make_prg/utils/io_utils.py:69]: why use a custom function for this when https://docs.python.org/3.6/library/tempfile.html exists?

File: make_prg/utils/misc.py
Comment (mbhall88) [inline make_prg/utils/misc.py:5]: `List`s of what? Do you want [`Any`](https://docs.python.org/3/library/typing.html#typing.Any)?

File: make_prg/utils/msa_aligner.py
Comment (mbhall88) [inline make_prg/utils/msa_aligner.py:91]: Again, why are we creating temporary directories? There's much less chance of things going wrong if we use `tempfile`?

File: make_prg/utils/seq_utils.py
Comment (mbhall88) [inline make_prg/utils/seq_utils.py:106]:
```suggestion
            assert sequence_is_composed_of_ACGT_only, f"Sequence ({sequence}) should be composed of ACTG only."
```

Comment (mbhall88) [reply make_prg/utils/seq_utils.py:106]: I thought we allowed some small number of ambiguities?

File: make_prg/utils/seq_utils.py
Comment (mbhall88) [inline make_prg/utils/seq_utils.py:168]: I think future devs will be thankful for a comment here that would indicate a) why this is bad, and b) why you think this should not be allowed to happen.  
Can you speculate as to what might have gone wrong?

**Review by mbhall88 — COMMENTED** [review]
I will let @bricoletc decide when this is approved. I just have some general questions and comments

File: make_prg/utils/io_utils.py
Comment (leoisl) [reply make_prg/utils/io_utils.py]: Yep, thanks for the correction!

File: make_prg/utils/io_utils.py
Comment (leoisl) [reply make_prg/utils/io_utils.py]: Thanks again, no idea why I kept this conditional

File: make_prg/utils/io_utils.py
Comment (leoisl) [reply make_prg/utils/io_utils.py]: You are right, now using `tempfile.mkdtemp()` (https://github.com/iqbal-lab-org/make_prg/pull/34/commits/ee0a383713c81630fc7505e6d901c07da83d76a6)

File: make_prg/utils/msa_aligner.py
Comment (leoisl) [reply make_prg/utils/msa_aligner.py]: Temporary dirs are needed for three reasons. 1. When we multiprocess, each process create 5 files per locus. To avoid filesystem bloat, we create one dir per locus, and put the 5 files there. These thousands of locus dirs are created inside the temporary dir. It would be too memory heavy to keep all this data in RAM, but maybe we could do some streaming process such that we don't need to write all these files, and later combine them. This might work, but it is a good amount of work just for not creating temp dirs/files, which is not a big issue. 2. `MAFFT` necessarily needs a temp dir (and possibly future aligners). 3. I wanted to control where the temp dir is created i.e. in the output folder, because I see many pipelines/runs failing because the space in the temp/scratch folder is full, especially in clusters. Sure, the standard is to create temp files in the standard temp dir (by default `/tmp`), and users can always override this default with `TMPDIR`, but I think creating temp dirs/files in the output dir will solve some headaches and issues in the future. This is my POV though, would be nice to know if you disagree with any of these points and propose an alternative solution.

Comment (mbhall88) [reply make_prg/utils/msa_aligner.py]: I'm not arguing that we don't need temporary directories - we absolutely do. I was just questioning why you are manually creating unique names etc. when `tempfile` handles this exact functionality - albeit thorough tested and supported. You can also tell `tempfile` *where* to create the temporary directories and files - it need not be `TMPDIR`.

Anyway, I'm not suggesting you make drastic changes. I was just flagging this as something that might be worth migrating to in the future and any new temporary file/dir functionality should maybe be handled by `tempfile`. Also, it would mean a lot less code to test.

File: make_prg/utils/seq_utils.py
Comment (leoisl) [reply make_prg/utils/seq_utils.py:106]: We do in the input, but not in the output, i.e. PRG. This is done just after expanding the alignments sequences, and when expanding, we transform non-ACGT bases into ACGT, and thus we should have just ACGT in the output. Also, if we encounter a base that is not ACGT, the binary PRG encoder, that produces input for `gramtools` is going to error out, as it just encodes ACGT (I think also both `gramtools` and `pandora` does not handle ambiguous bases in the PRG...).

File: make_prg/utils/misc.py
Comment (leoisl) [reply make_prg/utils/misc.py]: Thanks, improved the typing of this and the other function in `misc.py` (https://github.com/iqbal-lab-org/make_prg/pull/34/commits/6d7c6524d8f085538ba6db721c305eea388734d8)

File: make_prg/utils/msa_aligner.py
Comment (leoisl) [reply make_prg/utils/msa_aligner.py]: > I'm not arguing that we don't need temporary directories - we absolutely do. I was just questioning why you are manually creating unique names etc. when tempfile handles this exact functionality - albeit thorough tested and supported. You can also tell tempfile where to create the temporary directories and files - it need not be TMPDIR.

Yeah, just found this out when you linked the docs to tempfile, and implemented it here: https://github.com/iqbal-lab-org/make_prg/pull/34/commits/ee0a383713c81630fc7505e6d901c07da83d76a6

> Anyway, I'm not suggesting you make drastic changes. I was just flagging this as something that might be worth migrating to in the future and any new temporary file/dir functionality should maybe be handled by tempfile. Also, it would mean a lot less code to test.

Yep, has already implemented your suggestion here: https://github.com/iqbal-lab-org/make_prg/pull/34/commits/ee0a383713c81630fc7505e6d901c07da83d76a6

File: make_prg/utils/seq_utils.py
Comment (leoisl) [reply make_prg/utils/seq_utils.py]: Done here: https://github.com/iqbal-lab-org/make_prg/pull/34/commits/8a67d89778a238952037e702a49862be1f318f9b , does it look good? Also realise one of the asserts was not really required...

**Review by mbhall88 — APPROVED** [review]
As I mentioned before, I will leave it to Brice to approve this PR as he understands it much better than I do.

File: make_prg/utils/recursive_tree_drawer.py
Comment (bricoletc) [inline make_prg/utils/recursive_tree_drawer.py:31]: I'm guessing there is a reason that `_get_prg` fills in a list passed in as parameter rather than return a constructed list?

Comment (bricoletc) [reply make_prg/utils/recursive_tree_drawer.py:31]: (don't need to know it, just that i find the latter clearer)

File: make_prg/utils/seq_utils.py
Comment (bricoletc) [inline make_prg/utils/seq_utils.py:125]: I find it clearer to just do `for seq in remove_duplicates(sequences)`- but this is a minute point, can ignore

Comment (bricoletc) [reply make_prg/utils/seq_utils.py:125]: although because `remove_duplicates` returns a generator i think it's best not to assign a variable to it- the generator will be exhausted after traversal so the variable becomes useless

File: make_prg/utils/seq_utils.py
Comment (bricoletc) [inline make_prg/utils/seq_utils.py:159]: I remember @mbhall88  or martin saying BioPython's pairwise2 is v. slow- should we replace it with `edlib`?

File: make_prg/utils/seq_utils.py
Comment (bricoletc) [inline make_prg/utils/seq_utils.py:172]: when is it meaningful to align an empty sequence with a non-empty sequence? is there a case where, for a prg node(s) to be updated, this can actually happen?

File: make_prg/utils/seq_utils.py
Comment (bricoletc) [inline make_prg/utils/seq_utils.py:156]: I find this function's body's code very confusing, would this work instead
```
ref_is_empty = len(seq1) == 0
alt_is_empty = len(seq2) == 0
assert not (ref_is_empty and alt_is_empty), "Error"
if ref_is_empty:
    return GAP * len(seq2), seq2
elif alt_is_empty:
   return seq1, GAP * len(seq1)
else:
      alignment = pairwise2.align.globalms(
        seq1, seq2, match_score, mismatch_score, gap_open_score, gap_extend_score, one_alignment_only=True
    )
    return alignment[0].seqA, alignment[0].seqB
    
```

File: make_prg/utils/seq_utils.py
Comment (bricoletc) [inline make_prg/utils/seq_utils.py:178]: Just checking, where is this function used? I'm guessing in later PR code

File: tests/utils/test_msa_aligner.py
Comment (bricoletc) [inline tests/utils/test_msa_aligner.py:55]: may be worth commenting, or stating in function name, that code 1 signals failure, and that's whats being tested

File: tests/utils/test_msa_aligner.py
Comment (bricoletc) [inline tests/utils/test_msa_aligner.py:129]: what is actually being tested here? essentially that the execution of a fake mafft runs right?

File: tests/utils/test_msa_aligner.py
Comment (bricoletc) [inline tests/utils/test_msa_aligner.py:109]: this test is very complicated, and is essentially only testing that the function executes, without checking actual mafft outputs right?

i'm not sure i see this test's value

Comment (bricoletc) [reply tests/utils/test_msa_aligner.py:109]: sorry don't mean to be brutal it looks like it was hard to engineer

**Review by bricoletc — APPROVED** [review]
Great work, i didn't spot any bugs/mistakes.

Some code is entirely new and seems to prefigure upcoming update-related code, so I can't verify that it 'works', but I trust all the code i've seen!

I left comments that I think are all at your discretion

File: make_prg/utils/io_utils.py
Comment (leoisl) [reply make_prg/utils/io_utils.py:150]: Just for a slightly higher documentation purposes, so that we know exactly what `int(line_split[1])` means. We could instead remove this variable, and add `int(line_split[1])` to the total, and later, when re-reading the code, we could deduce what this last term means, but I prefer to just be explicit... More documentation never hurts!

File: make_prg/utils/io_utils.py
Comment (leoisl) [reply make_prg/utils/io_utils.py:99]: > I've seen you do this before, specifying the type of the member attribute, but is that really necessary since the type of the function param is already specified?

Yeah, this is just for more documentation purposes, like when I `ctrl` click on the attribute in `pycharm`, and get taken to this line, and immediately see the type of the attribute. Not much necessary though as, as you said, it is also in the function param... Just more documentation!

> Also, because the __init__ is just setting attributes, you could use a dataclass here

Indeed, I was not aware of `dataclass` at all! Will switch to it!

File: make_prg/utils/io_utils.py
Comment (bricoletc) [reply make_prg/utils/io_utils.py:150]: yeah it's a balance isn't it- keeping code terse but also documented, i'm happy with your answer ;)

File: make_prg/utils/io_utils.py
Comment (leoisl) [reply make_prg/utils/io_utils.py]: Switched to `dataclass` in https://github.com/iqbal-lab-org/make_prg/pull/34/commits/963185f294d4daf4d64a5f64abf7ec1c2f827618 . Also managed to switch class `ClusteringResult` to dataclass, as well as 3 other classes in the update code (not shown here). Just loved this std annotation, will make use of it a lot in the future, thanks!

File: make_prg/utils/io_utils.py
Comment (leoisl) [reply make_prg/utils/io_utils.py]: That is much cleaner, thanks, I indeed added the regexp pattern as a class variable (see https://github.com/iqbal-lab-org/make_prg/pull/34/commits/b12d38d60d132c097dcb34374f2913c9e52d736a)

File: make_prg/utils/io_utils.py
Comment (leoisl) [reply make_prg/utils/io_utils.py:150]: Yeah, it is a balance, not sure if there is a better or more correct way to do things in these cases, there is a lot of personal preferences too (maybe we should add an entry about this in `CONTRIBUTING.md`?)

File: make_prg/utils/io_utils.py
Comment (leoisl) [reply make_prg/utils/io_utils.py]: I see and agree! Fixed in https://github.com/iqbal-lab-org/make_prg/pull/34/commits/5781e3672ca903a521918b1ba92d5474a84b8131

File: make_prg/utils/io_utils.py
Comment (leoisl) [reply make_prg/utils/io_utils.py:171]: > so am I right that your new code allows constructing multiple loci in parallel right?

yes

> and then, it concatenates them?

yes

> Does it also make separate PRG files? I would need that for gramtools so that i can then add in the invariant bits of the genome in between
> specifically- i need the .bin files to be kept separate
> i think that's what you do in lines 185-190 below- sorry if i'm making you repeat yourself

Don't worry about it, first of all, I am making a bit messy for you to review this PR, as I made it so large that I had to break into modules, but modules depend on one another... so you have to review the PR with incomplete information, which is far from ideal. Although this code does not make separate PRGs (a single fasta with all PRG strings), we do build separate binary representation of PRGs, that can be used by `gramtools`. As a concrete example, this is the output of an integration test not yet PRed here: https://github.com/leoisl/make_prg/tree/update/tests/integration_tests/data/truth_output/several . If you download `several.prg.bin.zip`, you will see the separate `.bin` files. I think this means all is good for you and `gramtools`? Could you please confirm?

File: make_prg/utils/msa_aligner.py
Comment (leoisl) [reply make_prg/utils/msa_aligner.py:35]: yay! I really like the `get`/`set` pattern to ensure encapsulation, but rarely do it in python... Maybe sth to consider is if we add this pattern to `CONTRIBUTING.md`, i.e., all member variables are private and accessible only through set and get (or property methods). The annoying thing about it is that sometimes the code gets bloated with loads of `set` and `get` functions, so I am unsure about this...

File: make_prg/utils/seq_utils.py
Comment (leoisl) [reply make_prg/utils/seq_utils.py:125]: I do agree, leaving an exhausted generator in the code is error prone, implementing your suggestion

Comment (leoisl) [reply make_prg/utils/seq_utils.py]: Done in https://github.com/iqbal-lab-org/make_prg/pull/34/commits/f3bfa155a595c4a8f864d3d74abcf654cdf216b0

File: make_prg/utils/seq_utils.py
Comment (leoisl) [reply make_prg/utils/seq_utils.py:173]: Yes, for some specific indels. In rare cases, a PRG node can be an empty string, and if we find a denovo variant through it, we might align an empty sequence (the node's sequence) against a non-empty one (denovo variant). In a more common case, a denovo variant can be the complete deletion of all bases of a PRG node, then we would align the node's sequence with an empty string (this happens somewhat frequently when the PRG node is composed of a single base, e.g. in a SNP, and the denovo varant is the deletion of that base. Actually, in one of the previous iterations of the update feature, this was not implemented, and then it bugged out in real data due to these cases).

File: make_prg/utils/seq_utils.py
Comment (leoisl) [reply make_prg/utils/seq_utils.py:179]: Yes, in `RecursiveTreeNode.__init__()`, but I think it might be present in the current code also? It removes columns in the MSA that are full of gaps, like this second column here:

```
A-T
A-C
A-G
```

which can appear when we get sub-alignments from the full MSA.

File: tests/utils/test_msa_aligner.py
Comment (leoisl) [reply tests/utils/test_msa_aligner.py]: Done in https://github.com/iqbal-lab-org/make_prg/pull/34/commits/2ef3adb012a9214ae1f5207e1cf07b4a6a48c09f

File: tests/utils/test_msa_aligner.py
Comment (leoisl) [reply tests/utils/test_msa_aligner.py:109]: No worries, you were not brutal :p

It basically tests the whole [get_updated_alignment() function](https://github.com/leoisl/make_prg/blob/2e2ae14f74105e3d10f104ab53785fd28fba9309/make_prg/utils/msa_aligner.py#L88-L124), but by mocking heavy function calls, like creating temp dir for MAFFT to run in, creating MAFFT input files, running MAFFT itself, loading the updated alignment and cleaning up the run. This is indeed a complex function to test in the sense that it does lots of heavy stuff. The approach I followed here is mocking all these heavy stuff, as well as calls to other functions that were already unit tested, as I don't think we should require the user to have MAFFT installed in order to test this function, we can just mock it, that is what is usually done when you have a heavy dependency (e.g. a DB connection) during a test. But just mocking is not enough, as if we just mock the MAFFT call, but we invoke it with wrong parameters, the tests still would pass, whereas they should not. We need to make sure that the mocked calls received the parameters the way we expect, and that we know would work in a real execution, and this is asserted [by the end of the test case](https://github.com/leoisl/make_prg/blob/2e2ae14f74105e3d10f104ab53785fd28fba9309/tests/utils/test_msa_aligner.py#L132-L145)

About the value of this test case, IDK actually how to evaluate it, but I think it is worth keeping it...

File: tests/utils/test_msa_aligner.py
Comment (leoisl) [reply tests/utils/test_msa_aligner.py:129]: Answer in the response of the following comment

File: make_prg/utils/io_utils.py
Comment (mbhall88) [reply make_prg/utils/io_utils.py:150]: I don't think that needs to be in the contributing document. If others make contributions they can see the conventions in the code as it is.

File: make_prg/utils/io_utils.py
Comment (bricoletc) [reply make_prg/utils/io_utils.py:171]: perfect

File: make_prg/utils/seq_utils.py
Comment (bricoletc) [reply make_prg/utils/seq_utils.py:179]: i couldn't find it being used here but it's fine it made sense to me

File: tests/utils/test_msa_aligner.py
Comment (bricoletc) [reply tests/utils/test_msa_aligner.py:109]: yep works!

**Comment by bricoletc** [comment]
Merge in whenever you're ready @leoisl

File: make_prg/utils/seq_utils.py
Comment (leoisl) [reply make_prg/utils/seq_utils.py:156]: Yes, and it made things indeed much clearer, implemented in https://github.com/iqbal-lab-org/make_prg/pull/34/commits/97fe96eba4f294724c4d613718de50eb74f8d3c5 . Thank you!!!

**Pull request merged**
